fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java

```diff
@@ -95,7 +96,19 @@ public List<OptExpression> transform(OptExpression queryExpression, OptimizerCon
                 }
             }
         } else {
-            mvCandidateContexts = context.getCandidateMvs();
+            mvCandidateContexts.addAll(context.getCandidateMvs());
+        }
+        mvCandidateContexts.removeIf(x -> !x.prune(context, queryExpression));
+        MaterializationContext.RewriteOrdering ordering =
+                new MaterializationContext.RewriteOrdering(queryExpression, context.getColumnRefFactory());
+        mvCandidateContexts.sort(ordering);
```

Comment:
why not sort it in preprocessor? you can just sort once there, not sort when every mv rule applied.

Reply:
because the order is not fixed, but depends on the particular `OptExpr`